codec/src/main/java/io/netty/handler/codec/compression/Snappy.java

```diff
@@ -154,14 +197,36 @@ private static int hash(ByteBuf in, int index, int shift) {
     }
 
     /**
-     * Creates an appropriately sized hashtable for the given input size
+     * Returns a short[] to be used as a hashtable
      *
-     * @param inputSize The size of our input, ie. the number of bytes we need to encode
+     * @param hashTableSize the size for the hashtable
      * @return An appropriately sized empty hashtable
      */
-    private static short[] getHashTable(int inputSize) {
-        int hashTableSize = MathUtil.findNextPositivePowerOfTwo(inputSize);
-        return new short[Math.min(hashTableSize, MAX_HT_SIZE)];
+    private short[] getHashTable(int hashTableSize) {
+        if (reuseHashtable) {
+            return getHashTableFastThreadLocalArrayFill(hashTableSize);
+        } else {
+            return new short[hashTableSize];
+        }
+    }
+
+    /**
+     * Returns a short[] from a FastThreadLocal, zeroing for correctness
+     * creating a new one and resizing it if necessary
+     *
+     * @return An appropriately sized empty hashtable
+     * @param hashTableSize
+     */
+    public static short[] getHashTableFastThreadLocalArrayFill(int hashTableSize) {
+        short[] hashTable = HASH_TABLE.get();
+        if (hashTable == null || hashTable.length < hashTableSize) {
```

Comment:
Should we also limit the max length for "cached" tables ?

Reply:
@normanmaurer the max length is checked in the input for both kind of tables at line :116

```

int hashTableSize = MathUtil.findNextPositivePowerOfTwo(length);
hashTableSize = Math.min(hashTableSize, MAX_HT_SIZE);

```